### PR TITLE
Replacing \r\n line endings with \r line endings

### LINKED ISSUE #1091: Pasted text includes CRLF pairs where unneeded
Multiline text pasted from the clipboard includes CRLF pairs in all cases; this is inappropriate for "Unix-space" sessions, such as WSL.

# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.145]
Windows Terminal version (if applicable): 71e19cd82528d66a0a7867cbed85990cfc1685f1
```

# Steps to reproduce

Select multiline text in Terminal.
Copy it (via right-click, ostensibly)
Paste it (again via right-click)

# Expected behavior

When pasting into a Unix-space session -- such as WSL -- pasted text should have a reasonable set of line-ending characters.

# Actual behavior

Line endings are "doubled" on text paste to Unix-space sessions.

### PR DESCRIPTION
**Pull request by d-bingham**
## References

#1091 
#1094 
#2390 
#3314

## PR Checklist
* [x] Closes #1091
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Combination of the PRs #1094, #2390, and #3314, especially as discussed in #3314.

In short, this changes line endings from Windows-space \r\n to the more universal \r.

## Validation Steps Performed

Copied and pasted text into the terminal without the patch, line endings were doubled.
With the patch, line endings weren't doubled.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Does the `std::erase` solution not do it? Hmm

**Comment by d-bingham** [comment]
> Does the `std::erase` solution not do it? Hmm

It solves the problem if we're not concerned about stripping all \n characters; the solution here _only_ changes \r\n pairs.

**Comment by DHowett-MSFT** [comment]
ah, interesting. I wonder how common a degenerate clipboard is that contains `\r\n` and `\n` simultaneously, and how we should treat that. I also now wonder if all `\n` should become `\r` and all `\r\r` should become `\r`...

**Comment by d-bingham** [comment]
> ah, interesting. I wonder how common a degenerate clipboard is that contains `\r\n` and `\n` simultaneously, and how we should treat that. I also now wonder if all `\n` should become `\r` and all `\r\r` should become `\r`...

I tend to think it's warranted to err on the side of conservative character-stripping here.  `\r\r`, for example, is _probably_ legitimate -- for example, the sequence `Line 1\r\rLine 2` is double-spaced text in Unix-land, and I'm not sure there's any other reasonable interpretation of the sequence or reason to convert it to anything else.

An argument could be made to convert `\n` to `\r` universally (_after_ the `\r\n` to `\r` conversion, of course, to not just double up line-endings...), but I'm not sure it actually matters.

**Comment by d-bingham** [comment]
Oops, misread the comment a bit.  Anyway:

`\n` -> `\r`, then `\r\r` -> `\r` doesn't work because of the _possibility_ of getting valid Unix-land data.
`\r\n` ->`\r`, then `\n` -> `\r` _probably_ works? But if we're seeing a lone `\n`, we're not looking at standard Windows-land text, and it's not clear to me what assumptions can be made about it.

**Comment by DHowett-MSFT** [comment]
Sounds good to me. Thanks for digging in. :smile:

**Comment by fcharlie** [comment]
@d-bingham @DHowett-MSFT  When the string is long and multiple lines, can the following scheme reduce the memory copy?

```cpp
std::wstring ReplaceNewline(std::wstring_view str) {
  std::wstring s;
  s.reserve(str.size());
  while (!str.empty()) {
    auto pos = str.find(L"\r\n");
    if (pos == std::wstring::npos) {
      s.append(str);
      break;
    }
    s.append(str.substr(0, pos)).append(L"\n");
    str.remove_prefix(pos + 2);
  }
  return s;
}
```

**Comment by vvuk** [comment]
Err... apologies if I'm missing something, but:

> In short, this changes line endings from Windows-space \r\n to the more universal \r.

> for example, the sequence `Line 1\r\rLine 2` is double-spaced text in Unix-land

Unix-land (and pretty much everything non-Windows -- "more universal"?), uses `\n` line endings.  '\r' was only used as a line ending on MacOS 9 and earlier, and a few other much older places.

Shouldn't this patch/feature be changing `\r\n` to `\n`?

**Comment by DHowett-MSFT** [comment]
@vvuk you’re correct for textual data serialized to disk, but it appears (based on a couple popular terminal emulators available today) that the universal encoding used by terminal emulators for a newline is `U+000D CARRIAGE RETURN`, `\r`.

Windows Terminal’s backing pseudoterminal infrastructure transforms an incoming carriage return (_and_ an incoming line feed, which incidentally is the cause of the problem this PR fixed) into a key event encoding <kbd>Enter</kbd>.

That is handled on the other end in an application-specific manner.

**Comment by DHowett-MSFT** [comment]
I guess, in short, the Windows PTY always acts in `icrnl` mode. Judging by the behavior in putty and xterm, it makes the same assumption for data that transits the clipboard.

**Comment by vvuk** [comment]
Ahh, got it. I was indeed missing something. Thanks for the explanation :)

**Comment by noedelorme** [comment]
It is well more interesting to replace \r\n by \n. Indeed, thanks to this change we can paste several line in the terminal without executing them.
For example, I use Sagemath, and I always have to paste a whole python function but I can't because lines end with \r\n and the terminal execute thes lines after \r.
